File: solutions/cpp/generating-a-b-sqrt2-improved.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <functional>
#include <iostream>
#include <queue>
#include <random>
#include <unordered_set>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::function;
using std::hash;
using std::priority_queue;
using std::queue;
using std::random_device;
using std::uniform_int_distribution;
using std::unordered_set;
using std::vector;

// @include
struct ABSqrt2 {
  ABSqrt2(int a, int b) : a(a), b(b), val(a + b * sqrt(2)) {}

  // @exclude
  // Equal function for hash.
  bool operator==(const ABSqrt2& n) const { return a == n.a && b == n.b; }

  bool operator<(const ABSqrt2& that) const { return val > that.val; }
  // @include
  int a, b;
  double val;
};
// @exclude

// Hash function for ABSqrt2.
struct HashABSqrt2 {
  size_t operator()(const ABSqrt2& n) const {
    return hash<int>()(n.a) ^ hash<int>()(n.b);
  }
};
// @include

vector<ABSqrt2> GenerateFirstKABSqrt2(int k) {
  // Will store the first k numbers of the form a + b sqrt(2).
  vector<ABSqrt2> result;
  result.emplace_back(0, 0);
  int i = 0, j = 0;
  for (int n = 1; n < k; ++n) {
    ABSqrt2 result_i_plus_1(result[i].a + 1, result[i].b);
    ABSqrt2 result_j_plus_sqrt2(result[j].a, result[j].b + 1);
    if (result_i_plus_1.val < result_j_plus_sqrt2.val) {
      ++i;
      result.emplace_back(result_i_plus_1);
    } else if (result_i_plus_1.val > result_j_plus_sqrt2.val) {
      ++j;
      result.emplace_back(result_j_plus_sqrt2);
    } else {  // result_i_plus_1 == result_j_plus_sqrt2.
      ++i, ++j;
      result.emplace_back(result_i_plus_1);
    }
  }
  return result;
}
// @exclude

vector<ABSqrt2> Golden(int k) {
  priority_queue<ABSqrt2, vector<ABSqrt2>> min_heap;
  vector<ABSqrt2> smallest;
  unordered_set<ABSqrt2, HashABSqrt2> hash_table;

  // Initial for 0 + 0 * sqrt(2).
  min_heap.emplace(0, 0);
  hash_table.emplace(0, 0);

  while (smallest.size() < k) {
    ABSqrt2 s(min_heap.top());
    smallest.emplace_back(s);
    hash_table.erase(s);
    min_heap.pop();

    // Add the next two numbers derived from s.
    ABSqrt2 c1(s.a + 1, s.b), c2(s.a, s.b + 1);
    if (hash_table.emplace(c1).second) {
      min_heap.emplace(c1);
    }
    if (hash_table.emplace(c2).second) {
      min_heap.emplace(c2);
    }
  }
  return smallest;
}

static void SimpleTest() {
  vector<ABSqrt2> ans = GenerateFirstKABSqrt2(8);
  assert(0.0 == ans[0].val);
  assert(1.0 == ans[1].val);
  assert(sqrt(2.0) == ans[2].val);
  assert(2.0 == ans[3].val);
  assert(1.0 + sqrt(2.0) == ans[4].val);
  assert(2.0 * sqrt(2.0) == ans[5].val);
  assert(3.0 == ans[6].val);
  assert(2.0 + sqrt(2.0) == ans[7].val);
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  SimpleTest();
  for (int times = 0; times < 1000; ++times) {
    int k;
    if (argc == 2) {
      k = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      k = dis(gen);
    }
    vector<ABSqrt2> ans(GenerateFirstKABSqrt2(k));
    assert(ans.size() == k);
    for (size_t i = 0; i < ans.size(); ++i) {
      cout << ans[i].a << ' ' << ans[i].b << ' ' << ans[i].val << endl;
      if (i > 0) {
        assert(ans[i].val >= ans[i - 1].val);
      }
    }
    auto gold_res = Golden(k);
    assert(ans.size() == gold_res.size() &&
           equal(ans.begin(), ans.end(), gold_res.begin()));
  }
  return 0;
}

File: static/java/src/main/java/com/epi/GeneratingABSqrt2Improved.java
package com.epi;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Random;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;

public class GeneratingABSqrt2Improved {
  // @include
  public static class ABSqrt2 implements Comparable<ABSqrt2> {
    public int a, b;
    public double val;

    public ABSqrt2(int a, int b) {
      this.a = a;
      this.b = b;
      val = a + b * Math.sqrt(2);
    }
    // @exclude

    // Equal function for hash.
    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }

      ABSqrt2 that = (ABSqrt2)o;
      return a == that.a && b == that.b;
    }

    // Hash function for ABSqrt2.
    // clang-format off
    @Override
    public int hashCode() { return Objects.hash(a, b); }
    // clang-format on

    // clang-format off
    @Override
    public int compareTo(ABSqrt2 o) { return Double.compare(val, o.val); }
    // clang-format on
    // @include
  }

  public static List<ABSqrt2> generateFirstKABSqrt2(int k) {
    // Will store the first k numbers of the form a + b sqrt(2).
    List<ABSqrt2> result = new ArrayList<>();
    result.add(new ABSqrt2(0, 0));
    int i = 0, j = 0;
    for (int n = 1; n < k; ++n) {
      ABSqrt2 resultIPlus1 = new ABSqrt2(result.get(i).a + 1, result.get(i).b);
      ABSqrt2 resultJPlusSqrt2
          = new ABSqrt2(result.get(j).a, result.get(j).b + 1);
      if (resultIPlus1.val < resultJPlusSqrt2.val) {
        ++i;
        result.add(resultIPlus1);
      } else if (resultIPlus1.val > resultJPlusSqrt2.val) {
        ++j;
        result.add(resultJPlusSqrt2);
      } else { // resultIPlus1 == resultJPlusSqrt2.
        ++i;
        ++j;
        result.add(resultIPlus1);
      }
    }
    return result;
  }
  // @exclude

  public static List<ABSqrt2> golden(int k) {
    SortedSet<ABSqrt2> minHeap = new TreeSet<>();
    List<ABSqrt2> smallest = new ArrayList<>();
    Set<ABSqrt2> hash = new HashSet<>();

    // Initial for 0 + 0 * sqrt(2).
    minHeap.add(new ABSqrt2(0, 0));
    hash.add(new ABSqrt2(0, 0));

    while (smallest.size() < k) {
      ABSqrt2 s = minHeap.first();
      smallest.add(s);
      hash.remove(s);
      minHeap.remove(s);

      // Add the next two numbers derived from s.
      ABSqrt2 c1 = new ABSqrt2(s.a + 1, s.b);
      ABSqrt2 c2 = new ABSqrt2(s.a, s.b + 1);
      if (hash.add(c1)) {
        minHeap.add(c1);
      }
      if (hash.add(c2)) {
        minHeap.add(c2);
      }
    }
    return smallest;
  }

  private static void simpleTest() {
    List<ABSqrt2> ans = generateFirstKABSqrt2(8);
    assert(0.0 == ans.get(0).val);
    assert(1.0 == ans.get(1).val);
    assert(Math.sqrt(2.0) == ans.get(2).val);
    assert(2.0 == ans.get(3).val);
    assert(1.0 + Math.sqrt(2.0) == ans.get(4).val);
    assert(2.0 * Math.sqrt(2.0) == ans.get(5).val);
    assert(3.0 == ans.get(6).val);
    assert(2.0 + Math.sqrt(2.0) == ans.get(7).val);
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    for (int times = 0; times < 1000; ++times) {
      int k;
      if (args.length == 1) {
        k = Integer.parseInt(args[0]);
      } else {
        k = r.nextInt(10000) + 1;
      }
      List<ABSqrt2> ans = generateFirstKABSqrt2(k);
      assert(ans.size() == k);
      for (int i = 0; i < ans.size(); ++i) {
        System.out.println(ans.get(i).a + " " + ans.get(i).b + " "
                           + ans.get(i).val);
        if (i > 0) {
          assert(ans.get(i).val >= ans.get(i - 1).val);
        }
      }
      List<ABSqrt2> goldRes = golden(k);
      ans.equals(goldRes);
      /*
      for (int i = 0; i < k; ++i) {
        assert (ans.get(i).equals(goldRes.get(i)));
      }
      */
    }
  }
}
